### PR TITLE
[SPARK-12399] Display correct error message when accessing REST API with an unknown app Id

### PR DESCRIPTION
**Pull request by carsonwang**
I got an exception when accessing the below REST API with an unknown application Id.
`http://<server-url>:18080/api/v1/applications/xxx/jobs`
Instead of an exception, I expect an error message "no such app: xxx" which is a similar error message when I access `/api/v1/applications/xxx`

```
org.spark-project.guava.util.concurrent.UncheckedExecutionException: java.util.NoSuchElementException: no app with key xxx
    at org.spark-project.guava.cache.LocalCache$Segment.get(LocalCache.java:2263)
    at org.spark-project.guava.cache.LocalCache.get(LocalCache.java:4000)
    at org.spark-project.guava.cache.LocalCache.getOrLoad(LocalCache.java:4004)
    at org.spark-project.guava.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4874)
    at org.apache.spark.deploy.history.HistoryServer.getSparkUI(HistoryServer.scala:116)
    at org.apache.spark.status.api.v1.UIRoot$class.withSparkUI(ApiRootResource.scala:226)
    at org.apache.spark.deploy.history.HistoryServer.withSparkUI(HistoryServer.scala:46)
    at org.apache.spark.status.api.v1.ApiRootResource.getJobs(ApiRootResource.scala:66)
```

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala [hunk 1]
```diff
@@ -115,7 +115,17 @@ class HistoryServer(
...
-    Option(appCache.get(appKey))
+    try {
+      val ui = appCache.get(appKey)
+      Some(ui)
+    } catch {
+      case e: Exception => e.getCause() match {
+        case nsee: NoSuchElementException =>
+          None
+
+        case cause: Exception => throw cause
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: not needed

File: core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala [hunk 2]
```diff
@@ -115,7 +115,17 @@ class HistoryServer(
   }
 
   def getSparkUI(appKey: String): Option[SparkUI] = {
-    Option(appCache.get(appKey))
+    try {
+      val ui = appCache.get(appKey)
+      Some(ui)
+    } catch {
+      case e: Exception => e.getCause() match {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: `case NonFatal(e)`

**Comment by andrewor14** [comment]
@vanzin

**Comment by carsonwang** [comment]
retest this please

**Comment by carsonwang** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala [hunk 3]
```diff
@@ -115,7 +117,17 @@ class HistoryServer(
   }
 
   def getSparkUI(appKey: String): Option[SparkUI] = {
-    Option(appCache.get(appKey))
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: is it a better fix to do `appCache.getIfPresent` here?

Comment (carsonwang) [inline core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: `appCache.getIfPresent` returns null if there is no cached value for the appKey. But `appCache.get` will try to obtain that value from a `CacheLoader`, cache it and return it. So I think we still need use `appCache.get` here and handle the exception.

**Comment by vanzin** [comment]
LGTM, merging to master and 1.6. Thanks!

**Pull request closed without merging**